pkg/fuzzer/fuzzer.go

```diff
@@ -0,0 +1,307 @@
+)
+
+type Fuzzer struct {
+	Config    *Config
+	Corpus    *Corpus
+	NewInputs chan rpctype.Input
+
+	ctx    context.Context
+	mu     sync.Mutex
+	stats  map[string]uint64
+	rnd    *rand.Rand
+	target *prog.Target
+
+	choiceTable *choiceTableProxy
+
+	nextExec     *priorityQueue[*Request]
+	runningExecs map[*Request]time.Time
+	jobIDs       map[job]int64
+	nextJobID    int64
+
+	queuedCandidates atomic.Uint64
+	// If the source of candidates runs out of them, we risk
+	// generating too many needCandidate requests (one for
+	// each Config.MinCandidates). We prevent this with candidatesRequested.
+	candidatesRequested atomic.Bool
+}
+
+func NewFuzzer(ctx context.Context, cfg *Config, rnd *rand.Rand,
+	target *prog.Target) *Fuzzer {
+	f := &Fuzzer{
+		Config: cfg,
+		Corpus: newCorpus(),
+
+		ctx:          ctx,
+		stats:        map[string]uint64{},
+		rnd:          rnd,
+		target:       target,
+		choiceTable:  &choiceTableProxy{},
+		nextExec:     makePriorityQueue[*Request](),
+		runningExecs: map[*Request]time.Time{},
+		jobIDs:       map[job]int64{},
+	}
+	go func() {
+		<-ctx.Done()
+		close(f.NewInputs)
+	}()
+	go f.leakDetector()
+	f.updateChoiceTable()
+	return f
+}
+
+type Config struct {
+	Logf           func(level int, msg string, args ...interface{})
+	Coverage       bool
+	FaultInjection bool
+	Comparisons    bool
+	Collide        bool
+	EnabledCalls   map[*prog.Syscall]bool
+	NoMutateCalls  map[int]bool
+	LeakChecking   bool
+	FetchRawCover  bool
+	// If the number of queued candidates is less than MinCandidates,
+	// NeedCandidates is triggered.
+	MinCandidates  uint
+	NeedCandidates chan struct{}
+	NewInputs      chan rpctype.Input
+}
+
+type Request struct {
+	Prog         *prog.Prog
+	NeedCover    bool
+	NeedRawCover bool
+	NeedSignal   bool
+	NeedHints    bool
+	// Fields that are only relevant within pkg/fuzzer.
+	flags   ProgTypes
+	stat    string
+	result  *Result
+	resultC chan *Result
+}
+
+type Result struct {
+	Info *ipc.ProgInfo
+	Stop bool
+}
+
+func (fuzzer *Fuzzer) Done(req *Request, res *Result) {
+	// Triage individual calls.
+	// We do it before unblocking the waiting threads because
+	// it may result it concurrent modification of req.Prog.
+	if req.NeedSignal && res.Info != nil {
+		for call, info := range res.Info.Calls {
+			fuzzer.triageProgCall(req.Prog, &info, call, req.flags)
+		}
+		fuzzer.triageProgCall(req.Prog, &res.Info.Extra, -1, req.flags)
+	}
+	// Unblock threads that wait for the result.
+	req.result = res
+	if req.resultC != nil {
+		req.resultC <- res
+	}
+	// Update stats.
+	fuzzer.mu.Lock()
+	fuzzer.stats[req.stat]++
+	delete(fuzzer.runningExecs, req)
+	fuzzer.mu.Unlock()
+}
+
+func (fuzzer *Fuzzer) triageProgCall(p *prog.Prog, info *ipc.CallInfo, call int,
+	flags ProgTypes) {
+	prio := signalPrio(p, info, call)
+	if !fuzzer.Corpus.AddRawMaxSignal(info.Signal, prio) {
+		return
+	}
+	fuzzer.Logf(2, "found new signal in call %d in %s\n",
+		call, string(p.Serialize()))
```

Comment:
It's a pity we serialize it even that in most cases we don't log it.
Logging packages usually have V function that allows to check if a particular level is logged.
Will it be enough to log the short representation produced by Prog.String method? Then we could do:
```
	fuzzer.Logf(2, "found new signal in call %v in %v\n", call, p)
```